### ISSUE TITLE
Blackbox regression support

### ISSUE THREAD

**Issue by abigailgold** [Trusted-AI/adversarial-robustness-toolbox#1782]
**Is your feature request related to a problem? Please describe.**
At the moment regression model support is limited to a few frameworks (pytorch, sklearn, etc.)

**Describe the solution you'd like**
Similar to the BlackBoxClassifier it would be beneficial to also have a BlackBoxRegressor (for when the model object is not available or the correct framework-specific wrapper is not available).

**Comment by moohax** [comment]
The BlackboxClassifer class takes an arbitrary predict function. You should usually return a list of probabilities. 

But for regression models, I don't know ART has attacks for them. Perhaps integrating something like Optuna would be useful for regression problems - especially if you just need to move an output arbitrarily up or down.

**Comment by abigailgold** [comment]
Hi @moohax , there actually are a few attacks that support regression models, such as MembershipInferenceBlackBox, AttributeInferenceBlackBox, AttributeInferenceMembership.

I didn't really understand your comment about Optuna. Could you please elaborate on how it could be used in the context of ART attacks? Thanks.

**Comment by moohax** [comment]
@abigailgold , updated my comment above.

**Comment by abigailgold** [comment]
I'm not sure the above code would work though. You're telling the BlackboxClassifier that it is a binary classification task (with nb_classes=2) but regression_model.predict(x) is going to return floating point values.

**Comment by beat-buesser** [comment]
Hi @moohax and @abigailgold Thank you for the very interesting issue and discussion!

I see the advantage of posing an attack on regression as an attack on classification problem that it makes all classification-specific attacks become applicable.

I think many users might hesitate or not recognize the opportunity to use `BlackboxClassifier` for a regression model,  because of its name containing classifier.

I'm wondering if we can create a lightweight class `BlackboxRegressor` to guide the users of ART with the class name and include support for the upper and lower threshold values following @moohax 's example above?

**Comment by beat-buesser** [comment]
I also like the idea of having an evasion attack that just uses `optuna`, I think we should create a feature issue for it.

**Comment by abigailgold** [comment]
@moohax Yes, I understand now how you suggest to use BlackboxClassifier. I agree it could be used as is (just adapting the predict method accordingly), and also agree with @beat-buesser that it could be made this even easier for the user by supplying an API that would accept the threshold values and do this automatically for them. I also still think it would be beneficial to have a blackbox regression model to use with attacks that do support regressors. We will try to come up with a solution that satisfies both.
Regarding the use of optuna, I think this should be a separate feature request. @moohax is this something you would maybe be interested in contributing?

**Comment by moohax** [comment]
Though, I think the Optuna thing might be unpublished. I put together some research scripts to compare it against other evasion attacks. Should probably confirm our intuitions are correct 😁

**Comment by beat-buesser** [comment]
@moohax It's great to hear that you would be happy to contribute code! You would be most welcome! I could imagine that the performance of optuna depends on the loss function and norm of the constraints for the perturbations similar to the Carlini&Wagner attacks.

@abigailgold What do you think about using @moohax's proposal of `BlackboxModel` for the function that accepts classification and regression models and returns the specific estimator depending on the model type and arguments?
